BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+      &objectPathStart,
+      (int)objectLength, objectDisplayed ? objectDisplayed : "-"
+   );
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   free(completeFunctionName);
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
+   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
+}
+
+static void BacktraceFrame_delete(Object* object) {
+   BacktraceFrame* this = (BacktraceFrame*)object;
+   free(this->functionName);
+   free(this->demangleFunctionName);
+   free(this->objectPath);
+   free(this->objectName);
+   free(this);
+}
+
+static void BacktracePanel_setError(BacktracePanel* this, const char* error) {
+   assert(error);
+   assert(!this->error);
+
+   Panel* super = &this->super;
+   Panel_prune(super);
+
+   Vector* lines = this->super.items;
+   if (lines) {
+      Vector_delete(lines);
+   }
+
+   this->super.items = Vector_new(Class(ListItem), true, DEFAULT_SIZE);
+   Panel_set(super, 0, (Object*)ListItem_new(error, 0));
+}
+
+static void BacktracePanel_makePrintingHelper(const BacktracePanel* this, BacktracePanelPrintingHelper* printingHelper) {
+   Vector* lines = this->super.items;
+   size_t longestAddress = 0;
+
+   for (int i = 0; i < Vector_size(lines); i++) {
+      BacktraceFrame* frame = (BacktraceFrame*)Vector_get(lines, i);
+      frame->backtracePanel = this;
+
+      size_t digitOfOffsetFrame = strlen("+0x") + countDigit(frame->offset, 16);
+      if (frame->demangleFunctionName) {
+         size_t demangledFunctionNameLength = strlen(frame->demangleFunctionName) + digitOfOffsetFrame;
+         printingHelper->maxDemangledFuncNameLen = MAXIMUM(demangledFunctionNameLength, printingHelper->maxDemangledFuncNameLen);
+      }
+
+      if (frame->functionName) {
+         size_t functionNameLength = strlen(frame->functionName) + digitOfOffsetFrame;
+         printingHelper->maxFuncNameLen = MAXIMUM(functionNameLength, printingHelper->maxFuncNameLen);
+      }
+
+      if (frame->objectPath) {
+         size_t objectPathLength = strlen(frame->objectPath);
+         printingHelper->maxObjPathLen = MAXIMUM(objectPathLength, printingHelper->maxObjPathLen);
+      }
+
+      if (frame->objectName) {
+         size_t objectNameLength = strlen(frame->objectName);
+         printingHelper->maxObjNameLen = MAXIMUM(objectNameLength, printingHelper->maxObjNameLen);
+      }
+
+      longestAddress = MAXIMUM(frame->address, longestAddress);
+   }
+
+   size_t addressLength = MAX_HEX_ADDR_STR_LEN_32;
+   if (longestAddress > MAX_ADDR_32) {
+      addressLength = MAX_HEX_ADDR_STR_LEN_64;
```

Comment:
How about this for more-than-32-bit addresses?
```c
addressLength = strlen("0x") + countDigits(longestAddress, 16);
```

Reply:
For hexadecimal numbers, using CLZ might be the faster choice.

Reply:
```c
#ifndef ULLONG_WIDTH
#define ULLONG_WIDTH ((int)sizeof(unsigned long long) * CHAR_BIT)
#endif

addressLength = strlen("0x") + ((ULLONG_WIDTH / 4) - __builtin_clzll(longestAddress | 0x1) / 4);
```
The catch: If CLZ is not supported as a built-in, then we need an alternate implementation.

Reply:
I made this choice because I like how GDB prints the backtrace frame and especially the addresses (with a lot of zero).